managed/services/agents/service_info_broker.go

```diff
@@ -0,0 +1,235 @@
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+			TlsSkipVerify: agent.TLSSkipVerify,
+		}
+	case models.PostgreSQLServiceType:
+		tdp := agent.TemplateDelimiters(service)
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_POSTGRESQL_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+		}
+	case models.MongoDBServiceType:
+		tdp := agent.TemplateDelimiters(service)
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_MONGODB_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+			TextFiles: &agentpb.TextFiles{
+				Files:              agent.Files(),
+				TemplateLeftDelim:  tdp.Left,
+				TemplateRightDelim: tdp.Right,
+			},
+		}
+	case models.ProxySQLServiceType:
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_PROXYSQL_SERVICE,
+			Dsn:     agent.DSN(service, 2*time.Second, service.DatabaseName, nil),
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	case models.ExternalServiceType:
+		exporterURL, err := agent.ExporterURL(q)
+		if err != nil {
+			return nil, err
+		}
+
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_EXTERNAL_SERVICE,
+			Dsn:     exporterURL,
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	case models.HAProxyServiceType:
+		exporterURL, err := agent.ExporterURL(q)
+		if err != nil {
+			return nil, err
+		}
+
+		request = &agentpb.ServiceInfoRequest{
+			Type:    inventorypb.ServiceType_HAPROXY_SERVICE,
+			Dsn:     exporterURL,
+			Timeout: durationpb.New(3 * time.Second),
+		}
+	default:
+		return nil, errors.Errorf("unhandled Service type %s", service.ServiceType)
+	}
+	return request, nil
+}
+
+// GetInfoFromService sends a request to pmm-agent to query information from a service.
+func (c *ServiceInfoBroker) GetInfoFromService(ctx context.Context, q *reform.Querier, service *models.Service, agent *models.Agent) error {
+	l := logger.Get(ctx)
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > 4*time.Second {
+			l.Warnf("GetInfoFromService took %s.", dur)
+		}
+	}()
+
+	pmmAgentID := pointer.GetString(agent.PMMAgentID)
+	isSibSupported, err := isServiceInfoBrokerSupported(q, pmmAgentID)
+	if err != nil {
+		return err
+	}
+
+	if !isSibSupported {
+		return nil
+	}
+
+	// External exporters and haproxy do not support this functionality.
+	if service.ServiceType == models.ExternalServiceType || service.ServiceType == models.HAProxyServiceType {
+		return nil
+	}
+
+	pmmAgent, err := c.r.get(pmmAgentID)
+	if err != nil {
+		return err
+	}
```

Comment:
we can move these lines above https://github.com/percona/pmm/pull/2488/files#diff-2ad66e7e11412c93e69ccd469ad64a529a96ad48e6f62ed8daf554b966df83adR137 and in that case we can just pass pmmAgent to `isServiceInfoBrokerSupported` and we won't need to request pmm-agent from DB twice